pkg/network/self_sign_configuration.go

```diff
@@ -0,0 +1,130 @@
+package network
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/pkg/errors"
+
+	opv1alpha1 "github.com/kubevirt/cluster-network-addons-operator/pkg/apis/networkaddonsoperator/v1alpha1"
+)
+
+const (
+	caRotateIntervalDefault   = 7 * 24 * time.Hour // 7 days
+	caOverlapIntervalDefault  = 24 * time.Hour     // 1 day
+	certRotateIntervalDefault = 24 * time.Hour     // 1 day
+)
+
+// validateSelfSignConfiguration validates the following fields
+// - CARotateInterval
+// - CAOverlapInterval
+// - CertRotateInterval
+func validateSelfSignConfiguration(conf *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.SelfSignConfiguration == nil {
+		return []error{}
+	}
+
+	selfSignConfiguration := *conf.SelfSignConfiguration
+
+	errs := []error{}
+
+	caRotateInterval, err := parseCertificateKnob("caRotateInterval", selfSignConfiguration.CARotateInterval)
+	errs = appendOnError(errs, err)
+
+	caOverlapInterval, err := parseCertificateKnob("caOverlapInterval", selfSignConfiguration.CAOverlapInterval)
+	errs = appendOnError(errs, err)
+
+	certRotateInterval, err := parseCertificateKnob("certRotateInterval", selfSignConfiguration.CertRotateInterval)
+	errs = appendOnError(errs, err)
+
+	// If they cannot be parsed don't continue
+	if len(errs) > 0 {
+		return errs
+	}
+
+	err = validateGreaterThanZero("caRotateInterval", caRotateInterval)
+	errs = appendOnError(errs, err)
+
+	err = validateGreaterThanZero("caOverlapInterval", caOverlapInterval)
+	errs = appendOnError(errs, err)
+
+	err = validateGreaterThanZero("certRotateInterval", certRotateInterval)
+	errs = appendOnError(errs, err)
+
+	// If we have a zero value don't continue
+	if len(errs) > 0 {
+		return errs
+	}
+
+	err = validateGreaterThan("caRotateInterval", caRotateInterval, "caOverlapInterval", caOverlapInterval)
+	errs = appendOnError(errs, err)
+
+	err = validateGreaterThan("caRotateInterval", caRotateInterval, "certRotateInterval", certRotateInterval)
+	errs = appendOnError(errs, err)
+
+	return errs
+}
+
+func fillDefaultsSelfSignConfiguration(conf, previous *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.SelfSignConfiguration == nil || conf.SelfSignConfiguration.CARotateInterval == "" || conf.SelfSignConfiguration.CAOverlapInterval == "" || conf.SelfSignConfiguration.CertRotateInterval == "" {
+		if previous != nil && previous.SelfSignConfiguration != nil {
+			conf.SelfSignConfiguration = previous.SelfSignConfiguration
+			return []error{}
+		}
+
+		conf.SelfSignConfiguration = &opv1alpha1.SelfSignConfiguration{
```

Comment:
This may result in unexpected behavior. Let's say a user only configures the overlap, the config is accepted, but then they get completely different intervals.

We need to make sure that if somebody sets conf.SelfSignConfiguration, it always contains all 3 values. Would you please fail hard in parseCertificateKnob when the value is empty?

Reply:
We should validate _after_ setting the defaults if we do that.

Reply:
Why? When would it break if we keep it as is?

Reply:
Imaging that we have the following config
```yaml
caRotateInterval: ""
caOverlapInterval: "500h"
certRotateInterval: ""
```
Clearly `caOverlapInterval` is beyond `caRotateInterval` and that's wrong, so we either use defaults at validation or we set defaults and then validate.

Reply:
If somebody passes this config, it should fail validation with "you must provide all parameters". Would not that solve it?

Reply:
Ack that's good.